Enable addmm + GELU epilogue fusion via cuBLASLt

aakhundov opened the pull request:
Stack from [ghstack](https://github.com/ezyang/ghstack) (oldest at bottom):
* __->__ #103811

Summary:

Previously, addmm + GELU epilogue fusion was unconditionally disabled in `ATen/native/cuda/Blas.cpp` due to compilation and numerical issues in CUDA <= 11.4. This PR:

1. Enables addmm + GELU epilogue fusion for CUDA >= 11.8.

2. Restricts the usage of fused addmm epilogue to contiguous output (bugfix).

3. Extends unit tests with addmm epilogue fusion and GELU activation paths.

Test Plan:

$ python test/test_linalg.py -k test_addmm_relu -v

test_addmm_relu_cpu_bfloat16 (__main__.TestLinalgCPU.test_addmm_relu_cpu_bfloat16) ... ok
test_addmm_relu_cpu_float32 (__main__.TestLinalgCPU.test_addmm_relu_cpu_float32) ... ok
test_addmm_relu_cpu_float64 (__main__.TestLinalgCPU.test_addmm_relu_cpu_float64) ... ok
test_addmm_relu_cuda_bfloat16 (__main__.TestLinalgCUDA.test_addmm_relu_cuda_bfloat16) ... ok
test_addmm_relu_cuda_float32 (__main__.TestLinalgCUDA.test_addmm_relu_cuda_float32) ... ok
test_addmm_relu_cuda_float64 (__main__.TestLinalgCUDA.test_addmm_relu_cuda_float64) ... ok

$ python test/test_linalg.py -k test_addmm_gelu -v

test_addmm_gelu_cpu_bfloat16 (__main__.TestLinalgCPU.test_addmm_gelu_cpu_bfloat16) ... ok
test_addmm_gelu_cpu_float32 (__main__.TestLinalgCPU.test_addmm_gelu_cpu_float32) ... ok
test_addmm_gelu_cpu_float64 (__main__.TestLinalgCPU.test_addmm_gelu_cpu_float64) ... ok
test_addmm_gelu_cuda_bfloat16 (__main__.TestLinalgCUDA.test_addmm_gelu_cuda_bfloat16) ... ok
test_addmm_gelu_cuda_float32 (__main__.TestLinalgCUDA.test_addmm_gelu_cuda_float32) ... ok
test_addmm_gelu_cuda_float64 (__main__.TestLinalgCUDA.test_addmm_gelu_cuda_float64) ... ok

Reviewers: @eellison

Differential Revision: [D46829884](https://our.internmc.facebook.com/intern/diff/D46829884)

cc @ptrblck @csarofeen @xwang233

aakhundov committed:
Enable addmm + GELU epilogue fusion via cuBLASLt

Summary:

Previously, addmm + GELU epilogue fusion was unconditionally disabled in `ATen/native/cuda/Blas.cpp` due to compilation and numerical issues in CUDA <= 11.4. This PR:

1. Enables addmm + GELU epilogue fusion for CUDA >= 11.8.

2. Restricts the usage of fused addmm epilogue to contiguous output (bugfix).

3. Extends unit tests with addmm epilogue fusion and GELU activation paths.

Test Plan:

$ python test/test_linalg.py -k test_addmm_relu -v

test_addmm_relu_cpu_bfloat16 (__main__.TestLinalgCPU.test_addmm_relu_cpu_bfloat16) ... ok
test_addmm_relu_cpu_float32 (__main__.TestLinalgCPU.test_addmm_relu_cpu_float32) ... ok
test_addmm_relu_cpu_float64 (__main__.TestLinalgCPU.test_addmm_relu_cpu_float64) ... ok
test_addmm_relu_cuda_bfloat16 (__main__.TestLinalgCUDA.test_addmm_relu_cuda_bfloat16) ... ok
test_addmm_relu_cuda_float32 (__main__.TestLinalgCUDA.test_addmm_relu_cuda_float32) ... ok
test_addmm_relu_cuda_float64 (__main__.TestLinalgCUDA.test_addmm_relu_cuda_float64) ... ok

$ python test/test_linalg.py -k test_addmm_gelu -v

test_addmm_gelu_cpu_bfloat16 (__main__.TestLinalgCPU.test_addmm_gelu_cpu_bfloat16) ... ok
test_addmm_gelu_cpu_float32 (__main__.TestLinalgCPU.test_addmm_gelu_cpu_float32) ... ok
test_addmm_gelu_cpu_float64 (__main__.TestLinalgCPU.test_addmm_gelu_cpu_float64) ... ok
test_addmm_gelu_cuda_bfloat16 (__main__.TestLinalgCUDA.test_addmm_gelu_cuda_bfloat16) ... ok
test_addmm_gelu_cuda_float32 (__main__.TestLinalgCUDA.test_addmm_gelu_cuda_float32) ... ok
test_addmm_gelu_cuda_float64 (__main__.TestLinalgCUDA.test_addmm_gelu_cuda_float64) ... ok

Reviewers: @eellison

[ghstack-poisoned]

aakhundov commented:
/easycla

aakhundov commented:
@aakhundov has imported this pull request.  If you are a Meta employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D46829884).

IvanYashchuk commented on test/test_linalg.py:
In the future, we should move these tests to the `test_matmul_cuda.py` file. Other than that changes look good!

IvanYashchuk reviewed (approved):
My approval works only for the `test/test_linalg.py` file if merged with the GH bot.

aakhundov committed:
Update on "Enable addmm + GELU epilogue fusion via cuBLASLt"


Summary:

Previously, addmm + GELU epilogue fusion was unconditionally disabled in `ATen/native/cuda/Blas.cpp` due to compilation and numerical issues in CUDA <= 11.4. This PR:

1. Enables addmm + GELU epilogue fusion for CUDA >= 11.8.

2. Restricts the usage of fused addmm epilogue to contiguous output (bugfix).

3. Extends unit tests with addmm epilogue fusion and GELU activation paths.

Test Plan:

$ python test/test_linalg.py -k test_addmm_relu -v

test_addmm_relu_cpu_bfloat16 (__main__.TestLinalgCPU.test_addmm_relu_cpu_bfloat16) ... ok
test_addmm_relu_cpu_float32 (__main__.TestLinalgCPU.test_addmm_relu_cpu_float32) ... ok
test_addmm_relu_cpu_float64 (__main__.TestLinalgCPU.test_addmm_relu_cpu_float64) ... ok
test_addmm_relu_cuda_bfloat16 (__main__.TestLinalgCUDA.test_addmm_relu_cuda_bfloat16) ... ok
test_addmm_relu_cuda_float32 (__main__.TestLinalgCUDA.test_addmm_relu_cuda_float32) ... ok
test_addmm_relu_cuda_float64 (__main__.TestLinalgCUDA.test_addmm_relu_cuda_float64) ... ok

$ python test/test_linalg.py -k test_addmm_gelu -v

test_addmm_gelu_cpu_bfloat16 (__main__.TestLinalgCPU.test_addmm_gelu_cpu_bfloat16) ... ok
test_addmm_gelu_cpu_float32 (__main__.TestLinalgCPU.test_addmm_gelu_cpu_float32) ... ok
test_addmm_gelu_cpu_float64 (__main__.TestLinalgCPU.test_addmm_gelu_cpu_float64) ... ok
test_addmm_gelu_cuda_bfloat16 (__main__.TestLinalgCUDA.test_addmm_gelu_cuda_bfloat16) ... ok
test_addmm_gelu_cuda_float32 (__main__.TestLinalgCUDA.test_addmm_gelu_cuda_float32) ... ok
test_addmm_gelu_cuda_float64 (__main__.TestLinalgCUDA.test_addmm_gelu_cuda_float64) ... ok

Reviewers: eellison

Differential Revision: [D46829884](https://our.internmc.facebook.com/intern/diff/D46829884)

cc ptrblck csarofeen xwang233

[ghstack-poisoned]

aakhundov commented:
@aakhundov has imported this pull request.  If you are a Meta employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D46829884).

aakhundov committed:
Update on "Enable addmm + GELU epilogue fusion via cuBLASLt"


Summary:

Previously, addmm + GELU epilogue fusion was unconditionally disabled in `ATen/native/cuda/Blas.cpp` due to compilation and numerical issues in CUDA <= 11.4. This PR:

1. Enables addmm + GELU epilogue fusion for CUDA >= 11.8.

2. Restricts the usage of fused addmm epilogue to contiguous output (bugfix).

3. Extends unit tests with addmm epilogue fusion and GELU activation paths.

Test Plan:

$ python test/test_linalg.py -k test_addmm_relu -v

test_addmm_relu_cpu_bfloat16 (__main__.TestLinalgCPU.test_addmm_relu_cpu_bfloat16) ... ok
test_addmm_relu_cpu_float32 (__main__.TestLinalgCPU.test_addmm_relu_cpu_float32) ... ok
test_addmm_relu_cpu_float64 (__main__.TestLinalgCPU.test_addmm_relu_cpu_float64) ... ok
test_addmm_relu_cuda_bfloat16 (__main__.TestLinalgCUDA.test_addmm_relu_cuda_bfloat16) ... ok
test_addmm_relu_cuda_float32 (__main__.TestLinalgCUDA.test_addmm_relu_cuda_float32) ... ok
test_addmm_relu_cuda_float64 (__main__.TestLinalgCUDA.test_addmm_relu_cuda_float64) ... ok

$ python test/test_linalg.py -k test_addmm_gelu -v

test_addmm_gelu_cpu_bfloat16 (__main__.TestLinalgCPU.test_addmm_gelu_cpu_bfloat16) ... ok
test_addmm_gelu_cpu_float32 (__main__.TestLinalgCPU.test_addmm_gelu_cpu_float32) ... ok
test_addmm_gelu_cpu_float64 (__main__.TestLinalgCPU.test_addmm_gelu_cpu_float64) ... ok
test_addmm_gelu_cuda_bfloat16 (__main__.TestLinalgCUDA.test_addmm_gelu_cuda_bfloat16) ... ok
test_addmm_gelu_cuda_float32 (__main__.TestLinalgCUDA.test_addmm_gelu_cuda_float32) ... ok
test_addmm_gelu_cuda_float64 (__main__.TestLinalgCUDA.test_addmm_gelu_cuda_float64) ... ok

Reviewers: eellison

Differential Revision: [D46829884](https://our.internmc.facebook.com/intern/diff/D46829884)

cc ptrblck csarofeen xwang233

[ghstack-poisoned]

aakhundov commented:
@aakhundov has imported this pull request.  If you are a Meta employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D46829884).

aakhundov committed:
Update on "Enable addmm + GELU epilogue fusion via cuBLASLt"


Summary:

Previously, addmm + GELU epilogue fusion was unconditionally disabled in `ATen/native/cuda/Blas.cpp` due to compilation and numerical issues in CUDA <= 11.4. This PR:

1. Enables addmm + GELU epilogue fusion for CUDA >= 11.8.

2. Restricts the usage of fused addmm epilogue to contiguous output (bugfix).

3. Extends unit tests with addmm epilogue fusion and GELU activation paths.

Test Plan:

$ python test/test_linalg.py -k test_addmm_relu -v

test_addmm_relu_cpu_bfloat16 (__main__.TestLinalgCPU.test_addmm_relu_cpu_bfloat16) ... ok
test_addmm_relu_cpu_float32 (__main__.TestLinalgCPU.test_addmm_relu_cpu_float32) ... ok
test_addmm_relu_cpu_float64 (__main__.TestLinalgCPU.test_addmm_relu_cpu_float64) ... ok
test_addmm_relu_cuda_bfloat16 (__main__.TestLinalgCUDA.test_addmm_relu_cuda_bfloat16) ... ok
test_addmm_relu_cuda_float32 (__main__.TestLinalgCUDA.test_addmm_relu_cuda_float32) ... ok
test_addmm_relu_cuda_float64 (__main__.TestLinalgCUDA.test_addmm_relu_cuda_float64) ... ok

$ python test/test_linalg.py -k test_addmm_gelu -v

test_addmm_gelu_cpu_bfloat16 (__main__.TestLinalgCPU.test_addmm_gelu_cpu_bfloat16) ... ok
test_addmm_gelu_cpu_float32 (__main__.TestLinalgCPU.test_addmm_gelu_cpu_float32) ... ok
test_addmm_gelu_cpu_float64 (__main__.TestLinalgCPU.test_addmm_gelu_cpu_float64) ... ok
test_addmm_gelu_cuda_bfloat16 (__main__.TestLinalgCUDA.test_addmm_gelu_cuda_bfloat16) ... ok
test_addmm_gelu_cuda_float32 (__main__.TestLinalgCUDA.test_addmm_gelu_cuda_float32) ... ok
test_addmm_gelu_cuda_float64 (__main__.TestLinalgCUDA.test_addmm_gelu_cuda_float64) ... ok

Reviewers: eellison

Differential Revision: [D46829884](https://our.internmc.facebook.com/intern/diff/D46829884)

cc ptrblck csarofeen xwang233

[ghstack-poisoned]

aakhundov commented:
@aakhundov has imported this pull request.  If you are a Meta employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D46829884).

eellison commented on aten/src/ATen/native/cuda/Blas.cpp:
```diff
@@ -287,13 +287,13 @@ Tensor& addmm_out_cuda_impl(Tensor& result, const Tensor& self, const Tensor& ma
               result_->data_ptr<scalar_t>(),
               result_ld,
-#if 0
+#if defined(CUDA_VERSION) && CUDA_VERSION >= 11080
```
for anyone looking at this in the future: we're using 11.8 here because that's what our CI has test coverage for

eellison commented on test/test_linalg.py:
```diff
@@ -5534,9 +5569,18 @@ def test_addmm(self, device, dtype):
         self._test_addmm_impl(torch._addmm_activation, "relu", device, dtype)
 
+    @precisionOverride({torch.double: 1e-8, torch.float: 1e-4, torch.bfloat16: 0.6,
+                        torch.half: 1e-1, torch.cfloat: 1e-4, torch.cdouble: 1e-8})
```
i'm surprised the torch.half bounds are so high here, but this is a prior issue, no need to fix on your pr

aakhundov replied:
My understanding is that these tests are not run on `torch.half` (only on `float32`, `float64`, and `bfloat16` on CUDA). So maybe it's a copy / paste artifact.

eellison reviewed (approved):
nice !!

aakhundov commented:
@pytorchbot merge

aakhundov commented:
@pytorchbot label "topic: not user facing"

aakhundov commented:
@pytorchbot merge

huydhn commented:
cc @aakhundov I suspect that this change is causing failures on periodic Windows CUDA job https://hud.pytorch.org/pytorch/pytorch/commit/1c79003b3c13c7bc47e5796e4451d6565121f3a0.  Could you help take a look into the issue? May be relax the tolerance on Windows?

```
2023-06-22T01:54:28.9232875Z =================================== RERUNS ====================================
2023-06-22T01:54:28.9233147Z _________________ TestLinalgCUDA.test_addmm_gelu_cuda_float32 _________________
2023-06-22T01:54:28.9233410Z Traceback (most recent call last):
2023-06-22T01:54:28.9233730Z   File "C:\actions-runner\_work\pytorch\pytorch\test\test_linalg.py", line 5582, in test_addmm_gelu
2023-06-22T01:54:28.9234092Z     self._test_addmm_impl(torch._addmm_activation, "gelu", device, dtype)
2023-06-22T01:54:28.9234464Z   File "C:\actions-runner\_work\pytorch\pytorch\test\test_linalg.py", line 5526, in _test_addmm_impl
2023-06-22T01:54:28.9234830Z     self._test_addmm_addmv(func, V, m1, m2, beta=1, activation=activation)
2023-06-22T01:54:28.9235268Z   File "C:\actions-runner\_work\pytorch\pytorch\test\test_linalg.py", line 5447, in _test_addmm_addmv
2023-06-22T01:54:28.9235818Z     self.assertEqual(res1, res3)
2023-06-22T01:54:28.9236206Z   File "C:\actions-runner\_work\pytorch\pytorch\build\win_tmp\build\torch\testing\_internal\common_utils.py", line 3100, in assertEqual
2023-06-22T01:54:28.9236552Z     raise error_metas[0].to_error(
2023-06-22T01:54:28.9236787Z AssertionError: Tensor-likes are not close!
2023-06-22T01:54:28.9236938Z 
2023-06-22T01:54:28.9237027Z Mismatched elements: 67 / 250 (26.8%)
2023-06-22T01:54:28.9237294Z Greatest absolute difference: 0.00047320593148469925 at index (2, 5) (up to 0.0001 allowed)
2023-06-22T01:54:28.9237609Z Greatest relative difference: 0.3691028356552124 at index (6, 22) (up to 1.3e-06 allowed)
```

aakhundov commented:
@huydhn Thanks for the heads-up! From the error, it seems that the GELU epilogue fusion is not available on Windows (or it doesn't use tanh approximation). In any case, we should be able to fix this by adding `not IS_LINUX` at this line: https://github.com/pytorch/pytorch/blob/main/test/test_linalg.py#L5419 . I'll send a PR shortly.

Btw, the failing Windows CUDA job didn't show up in the CI of the PR. Is this expected? Is there a way to trigger the Windows CUDA job manually in the PR, before merging? Thanks!

aakhundov commented:
@huydhn Submitted a fix in #104031. Would be great if we could run the Windows CUDA job on this PR to make sure it remedies the issue. Thanks!

huydhn commented:
> Btw, the failing Windows CUDA job didn't show up in the CI of the PR. Is this expected? Is there a way to trigger the Windows CUDA job manually in the PR, before merging? Thanks!

Yes, you can add `ciflow/periodic` to your PR to trigger Windows CUDA jobs.  As the name implies, the job only runs periodically because we only have a limited number of Windows CUDA runners.  When failure like this happens, it's ok to follow up with a fix.

I have add `ciflow/periodic` onto https://github.com/pytorch/pytorch/pull/104031, so it should be running there now.
